lib/Dialect/FIRRTL/Analysis/DetectCombLoops.cpp

```diff
@@ -0,0 +1,228 @@
+
+  /// DFS stack used to maintain the ordering.
+  SmallVector<Value> valueStack;
+
+  /// A reference to the global combinational paths map of the circuit.
+  DenseMap<Operation *, InputOutputPaths> &combPathsMap;
+
+  /// Combinational paths of this module.
+  InputOutputPaths combPaths;
+
+  /// Store a module argument when it is the root node of the current DFS
+  /// traversing. This is introduced to detect and record combinational paths
+  /// between input and output arguments.
+  BlockArgument currentArg = nullptr;
+};
+} // namespace
+
+/// Detect combinational loops in the module.
+void Detector::detect() {
+  // We first start the DFS from all module arguments. This intends to detect
+  // all combinational paths between input and output arguments of the module.
+  for (auto arg : module.getArguments()) {
+    currentArg = arg;
+    if (!arg.getType().isa<ClockType>())
+      DFSVisitValue(arg);
+  }
+
+  // Record the combinational paths and set current argument to null.
+  combPathsMap[module] = combPaths;
+  currentArg = nullptr;
+
+  // As FModuleOp contains an normal SSA region rather than a graph region,
+  // every combinational loop must have a ConnectOp on its path.
+  for (auto connect : module.getOps<ConnectOp>())
+    DFSVisitValue(connect.dest());
+}
+
+/// DFS visit a value. Return the minimum DFS index in the SCC.
+int64_t Detector::DFSVisitValue(Value value) {
+  if (valueIndexMap.count(value))
+    return valueIndexMap[value];
+
+  // Record the minimum index in the SCC.
+  auto minIndex = index;
+
+  // Get the index of the current value and increase the global index.
+  valueIndexMap[value] = index++;
+
+  // Early return when meeting a register.
+  auto definingOp = value.getDefiningOp();
+  if (definingOp && isa<RegOp, RegResetOp>(definingOp))
+    return minIndex;
+
+  // If the current value is a module argument and not equal to `currentArg`, we
+  // should record this path as a combinational path.
+  if (!definingOp && currentArg && currentArg != value) {
+    auto outputIndex = BlockArgument(value.getImpl()).getArgNumber();
+    combPaths[currentArg.getArgNumber()].push_back(outputIndex);
+  }
+
+  auto currentStackSize = valueStack.size();
+  valueStack.push_back(value);
+
+  // Collect childs of the current value.
+  SmallVector<Value> childs;
+  for (auto user : value.getUsers()) {
+    llvm::append_range(childs, user->getResults());
+    // If the current value is the `src` of a connect operation, the `dest`
+    // value should be considered as a child.
+    if (auto connect = dyn_cast<ConnectOp>(user)) {
+      if (value != connect.dest())
+        childs.push_back(connect.dest());
+    }
+  }
+
+  // Handle two special operations that may introduce extra childs.
+  if (definingOp) {
+    // Handle instance operation.
+    if (auto instance = dyn_cast<InstanceOp>(definingOp)) {
+      if (auto referencedModule =
+              dyn_cast<FModuleOp>(instance.getReferencedModule())) {
```

Comment:
Plz use Andrew's fancy CallGraph analysis which will make this O(1) instead of O(n)